internal/catalog/internal/controllers/workloadhealth/controller.go

```diff
@@ -0,0 +1,210 @@
+	}
+
+	return controller.ForType(types.WorkloadType).
+		WithWatch(types.HealthStatusType, controller.MapOwnerFiltered(types.WorkloadType)).
+		WithWatch(types.NodeType, nodeMap.MapNodeToWorkloads).
+		WithReconciler(&workloadHealthReconciler{nodeMap: nodeMap})
+}
+
+type workloadHealthReconciler struct {
+	nodeMap NodeMapper
+}
+
+func (r *workloadHealthReconciler) Reconcile(ctx context.Context, rt controller.Runtime, req controller.Request) error {
+	// read the workload
+	rsp, err := rt.Client.Read(ctx, &pbresource.ReadRequest{Id: req.ID})
+	switch {
+	case status.Code(err) == codes.NotFound:
+		r.nodeMap.RemoveWorkloadTracking(req.ID)
+		return nil
+	case err != nil:
+		return err
+	}
+
+	res := rsp.Resource
+	var workload pbcatalog.Workload
+	if err := res.Data.UnmarshalTo(&workload); err != nil {
+		// This should be impossible and will not be exercised in tests. Various
+		// type validations on admission ensure that all Workloads would
+		// be unmarshallable in this way.
+		return err
+	}
+
+	nodeHealth := pbcatalog.Health_HEALTH_PASSING
+	if workload.NodeName != "" {
+		nodeID := r.nodeMap.NodeIDFromWorkload(res, &workload)
+		r.nodeMap.TrackWorkload(res.Id, nodeID)
+		nodeHealth, err = getNodeHealth(ctx, rt, nodeID)
+		if err != nil {
+			return err
+		}
+	} else {
+		// the node association may be been removed so stop tracking it.
+		r.nodeMap.RemoveWorkloadTracking(res.Id)
+	}
+
+	workloadHealth, err := getWorkloadHealth(ctx, rt, req.ID)
+	if err != nil {
+		// This should be impossible under normal operations and will not be exercised
+		// within the unit tests. This can only fail if the resource service fails
+		// or allows admission of invalid health statuses.
+		return err
+	}
+
+	health := nodeHealth
+	if workloadHealth > health {
+		health = workloadHealth
+	}
+
+	statusState := pbresource.Condition_STATE_TRUE
+	if health != pbcatalog.Health_HEALTH_PASSING {
+		statusState = pbresource.Condition_STATE_FALSE
+	}
+
+	message := WorkloadHealthyMessage
+	if workload.NodeName != "" {
+		message = NodeAndWorkloadHealthyMessage
+	}
+	switch {
+	case workloadHealth != pbcatalog.Health_HEALTH_PASSING && nodeHealth != pbcatalog.Health_HEALTH_PASSING:
+		message = NodeAndWorkloadUnhealthyMessage
+	case workloadHealth != pbcatalog.Health_HEALTH_PASSING:
+		message = WorkloadUnhealthyMessage
+	case nodeHealth != pbcatalog.Health_HEALTH_PASSING:
+		message = nodehealth.NodeUnhealthyMessage
+	}
+
+	newStatus := &pbresource.Status{
+		ObservedGeneration: res.Generation,
+		Conditions: []*pbresource.Condition{
+			{
+				Type:    StatusConditionHealthy,
+				State:   statusState,
+				Reason:  health.String(),
+				Message: message,
+			},
+		},
+	}
+
+	if resource.EqualStatus(res.Status[StatusKey], newStatus, false) {
+		return nil
+	}
+
+	_, err = rt.Client.WriteStatus(ctx, &pbresource.WriteStatusRequest{
+		Id:     res.Id,
+		Key:    StatusKey,
+		Status: newStatus,
+	})
```

Comment:
The propagation of this logic and the tests for it makes me feel like it should be in a helper because every controller would need to do the same thing.

Reply:
@boxofrad I agree with Iryna. Where do you think this logic should live. It could be a method on the controller.Runtime or it could live as a helper within the resource package.

Reply:
Specifically, do you mean calling `WriteStatus` only if it's changed?

Runtime might work well. I'm imagining the signature will be:

```Go
func (r Runtime) WriteStatus(ctx context.Context, res *pbresource.Resource, key string, status *pbresource.Status) error {
    // ...
}
```

Is that roughly what you have in mind?